docs/design/llvm_sparse_runtime.md

```diff
@@ -0,0 +1,458 @@
+
+```cpp
+void ListManager::touch_chunk(int chunk_id) {
+  taichi_assert_runtime(runtime, chunk_id < max_num_chunks,
+                        "List manager out of chunks.");
+  if (!chunks[chunk_id]) {
+    locked_task(&lock, [&] {
+      // may have been allocated during lock contention
+      if (!chunks[chunk_id]) {
+        grid_memfence();
+        auto chunk_ptr = runtime->request_allocate_aligned(
+            max_num_elements_per_chunk * element_size, 4096);
+        atomic_exchange_u64((u64 *)&chunks[chunk_id], (u64)chunk_ptr);
+      }
+    });
+  }
+}
+```
+
+By now, you should be quite familiar with this double-checked locking pattern. One thing noteworthy is that the fundamental memory allocation function is provided by [`LLVMRuntime::request_allocate_aligned()`](https://github.com/taichi-dev/taichi/blob/172cab8a57fcfc2d766fe2b7cd40af669dadf326/taichi/runtime/llvm/runtime.cpp#L846-L866).
+
+[`get_element_ptr()`](https://github.com/taichi-dev/taichi/blob/172cab8a57fcfc2d766fe2b7cd40af669dadf326/taichi/runtime/llvm/runtime.cpp#L480-L483) illustrates the way to lookup the address of an element from a given index:
+
+```cpp
+Ptr get_element_ptr(i32 i) {
+  return chunks[i >> log2chunk_num_elements] +                       // chunk base
+          element_size * (i & ((1 << log2chunk_num_elements) - 1));  // slot within the chunk
+}
+```
+
+There is also a reverse operation, [`ptr2index()`](https://github.com/taichi-dev/taichi/blob/172cab8a57fcfc2d766fe2b7cd40af669dadf326/taichi/runtime/llvm/runtime.cpp#L499-L509). It iterates over each chunk to see if the passed-in `ptr` falls into that memory range.
+
+With these primitive helpers, [`allocate()`](https://github.com/taichi-dev/taichi/blob/172cab8a57fcfc2d766fe2b7cd40af669dadf326/taichi/runtime/llvm/runtime.cpp#L1591-L1594) is trivial to implement:
+
+```cpp
+Ptr ListManager::allocate() {
+  auto i = reserve_new_element();
+  return get_element_ptr(i);
+}
+```
+
+So is [`append()`](https://github.com/taichi-dev/taichi/blob/172cab8a57fcfc2d766fe2b7cd40af669dadf326/taichi/runtime/llvm/runtime.cpp#L1586-L1589):
+
+```cpp
+void ListManager::append(void *data_ptr) {
+  auto ptr = allocate();
+  std::memcpy(ptr, data_ptr, element_size);
+}
+```
+
+[`clear()`](https://github.com/taichi-dev/taichi/blob/172cab8a57fcfc2d766fe2b7cd40af669dadf326/taichi/runtime/llvm/runtime.cpp#L472-L474) simply resets `num_elements` to zero, without doing anything to the list contents.
+
+## GC
+
+The GC process for a given SNode is divided into three stages:
+
+1. [`gc_parallel_0`](https://github.com/taichi-dev/taichi/blob/172cab8a57fcfc2d766fe2b7cd40af669dadf326/taichi/runtime/llvm/runtime.cpp#L1600-L1626): Moves the remaining, unused indices in `free_list` to its head.
+
+The code differentiates between the cases where destination and source ranges overlap or not:
```

Comment:
```suggestion
Normally this can be done via a simple `for` loop that copies data. However, we need to run it in parallel on GPUs and the data source and destination may overlap, so special care is needed.
Therefore, the code differentiates between the cases where destination and source ranges overlap or not:
```